runtime/src/libc/syscall.c

```diff
@@ -318,25 +320,20 @@ wasm_mremap(int32_t offset, int32_t old_size, int32_t new_size, int32_t flags)
 	if (new_size <= old_size) return offset;
 
 	// If at end of linear memory, just expand and return same address
-	if (offset + old_size == local_sandbox_context_cache.memory.size) {
-		int32_t amount_to_expand  = new_size - old_size;
-		int32_t pages_to_allocate = amount_to_expand / WASM_PAGE_SIZE;
-		if (amount_to_expand % WASM_PAGE_SIZE > 0) pages_to_allocate++;
-		for (int i = 0; i < pages_to_allocate; i++) expand_memory();
-
+	if (offset + old_size == current_wasm_module_instance.memory->size) {
+		int32_t amount_to_expand = new_size - old_size;
+		wasm_memory_expand(current_wasm_module_instance.memory, amount_to_expand);
 		return offset;
 	}
 
 	// Otherwise allocate at end of address space and copy
-	int32_t pages_to_allocate = new_size / WASM_PAGE_SIZE;
-	if (new_size % WASM_PAGE_SIZE > 0) pages_to_allocate++;
-	int32_t new_offset = local_sandbox_context_cache.memory.size;
-	for (int i = 0; i < pages_to_allocate; i++) expand_memory();
+	int32_t new_offset = current_wasm_module_instance.memory->size;
+	wasm_memory_expand(current_wasm_module_instance.memory, new_size);
 
 	// Get pointer of old offset and pointer of new offset
-	char *linear_mem = local_sandbox_context_cache.memory.start;
-	char *src        = &linear_mem[offset];
-	char *dest       = &linear_mem[new_offset];
+	uint8_t *linear_mem = current_wasm_module_instance.memory->data;
```

Comment:
Can't tell if the additional level of indirection is on the fastpath (memory access) or not. I'll assume not, unless told otherwise ;-)

Reply:
I'll revisit this. I think I can restructure things to flatten them out as long as I drop flexible array members entirely and replace them with pointers.